File: config/src/main/scala/net/kemitix/thorp/config/ParseConfigLines.scala
package net.kemitix.thorp.config

import java.nio.file.Paths
import java.util.regex.Pattern

import net.kemitix.thorp.config.ConfigOption._
import zio.UIO

trait ParseConfigLines {

  private val pattern = "^\\s*(?<key>\\S*)\\s*=\\s*(?<value>\\S*)\\s*$"
  private val format  = Pattern.compile(pattern)

  def parseLines(lines: Seq[String]): UIO[ConfigOptions] =
    UIO(ConfigOptions(lines.flatMap(parseLine).toList))

  private def parseLine(str: String) =
    format.matcher(str) match {
      case m if m.matches => parseKeyValue(m.group("key"), m.group("value"))
      case _              => List.empty
    }

  private def parseKeyValue(
      key: String,
      value: String
  ): List[ConfigOption] =
    key.toLowerCase match {
      case "source"  => List(Source(Paths.get(value)))
      case "bucket"  => List(Bucket(value))
      case "prefix"  => List(Prefix(value))
      case "include" => List(Include(value))
      case "exclude" => List(Exclude(value))
      case "debug"   => if (truthy(value)) List(Debug()) else List.empty
      case _         => List.empty
    }

  private def truthy(value: String): Boolean =
    value.toLowerCase match {
      case "true"    => true
      case "yes"     => true
      case "enabled" => true
      case _         => false
    }

}

object ParseConfigLines extends ParseConfigLines

File: config/src/main/java/net/kemitix/thorp/config/ParseConfigLines.java
package net.kemitix.thorp.config;

import java.nio.file.Paths;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ParseConfigLines {
    private static final String pattern = "^\\s*(?<key>\\S*)\\s*=\\s*(?<value>\\S*)\\s*$";
    private static final Pattern format = Pattern.compile(pattern);

    ConfigOptions parseLines(List<String> lines) {
        return ConfigOptions.create(
                lines.stream()
                        .flatMap(this::parseLine)
                        .collect(Collectors.toList()));
    }

    private Stream<ConfigOption> parseLine(String str) {
        Matcher m = format.matcher(str);
        if (m.matches()) {
            return parseKeyValue(m.group("key"), m.group("value"));
        }
        return Stream.empty();
    }

    private Stream<ConfigOption> parseKeyValue(String key, String value) {
        switch (key.toLowerCase()) {
            case "parallel":
                return Stream.of(ConfigOption.parallel(Integer.parseInt(value)));
            case "source":
                return Stream.of(ConfigOption.source(Paths.get(value)));
            case "bucket":
                return Stream.of(ConfigOption.bucket(value));
            case "prefix":
                return Stream.of(ConfigOption.prefix(value));
            case "include":
                return Stream.of(ConfigOption.include(value));
            case "exclude":
                return Stream.of(ConfigOption.exclude(value));
            case "debug":
                if (truthy(value))
                    return Stream.of(ConfigOption.debug());
                // fall through to default
            default:
                return Stream.empty();
        }
    }

    private boolean truthy(String value) {
        switch (value.toLowerCase()) {
            case "true":
            case "yes":
            case "enabled":
                return true;
            default:
                return false;
        }
    }
}
